node/silkworm/stagedsync/stage_interhashes/trie_cursor.hpp

```diff
@@ -0,0 +1,137 @@
+    struct move_operation_result {
+        bool skip_state{false};        // Whether the node can be used as is without need to recompute root hash
+        std::optional<Bytes> key{};    // Nibbled key of node
+        std::optional<Bytes> hash{};   // Hash of node
+        bool children_in_trie{false};  // Whether there are children in trie
+        std::optional<Bytes> first_uncovered{};  // First uncovered prefix to be processed by higher loop
+    };
+
+    //! \brief Acquires the prefix and position the cursor to the first occurrence
+    [[nodiscard]] move_operation_result to_prefix(ByteView prefix);
+
+    //! \brief Moves the cursor to next relevant position
+    [[nodiscard]] move_operation_result to_next();
+
+  private:
+    struct SubNode {
+        ByteView key{};            // Key retrieved from db (if any) Is nibbled
+        ByteView value{};          // Value retrieved from db (if any)
+        uint16_t state_mask{0};    // One bit set for every child_id which points to a hashed state
+        uint16_t tree_mask{0};     // One bit set for every child_id which has a child in trie
+        uint16_t hash_mask{0};     // One bit set for every child_id which has a valid hash
+        ByteView root_hash{};      // Root Hash
+        ByteView hashes{};         // Child nodes hashes data
+        size_t hashes_count{0};    // Actual number of hashes (excluding root hash)
+        int8_t child_id{-1};       // Current child being inspected in this node (aka nibble)
+        int8_t max_child_id{0xf};  // Max child of this node
+        int8_t hash_id{-1};        // Index of hash to be retrieved
+        bool deleted{false};       // Whether already deleted (in collector)
+
+        [[nodiscard]] bool has_tree() const noexcept;   // Whether current child_id has bit set in tree mask
+        [[nodiscard]] bool has_hash() const noexcept;   // Whether current child_id has bit set in hash mask
+        [[nodiscard]] bool has_state() const noexcept;  // Whether current child_id has bit set in state mask
+
+        void reset();                                   // Resets node to default values
+        void parse(ByteView k, ByteView v);             // Parses node data contents from db (may throw)
+        [[nodiscard]] Bytes full_key() const noexcept;  // Returns full key to child node (i.e. key + child_id)
+        [[nodiscard]] Bytes hash() const;               // Returns hash of child node (i.e. key + child_id)
+    };
+
+    uint32_t level_{0};                      // Depth level in sub_nodes_
+    bool end_of_tree_{false};                // Protects from to_next() beyond end of tree
+    Bytes curr_key_{};                       // Latest key returned on a valid hash
+    Bytes prev_key_{};                       // Same as curr_key_ but for previous cycle
+    bool skip_state_{true};                  // Whether account(s) state scan can be skipped
+    std::array<SubNode, 64> sub_nodes_{{}};  // Collection of sub-nodes being unrolled
```

Comment:
Why 64?

Reply:
Oh ... thank you for spotting.
It's a legacy of Erigon's reasoning where "level_" was set on behalf of nibbled key len.
Actually 32 is the theoretical max value is 32 (1 for root + 31 of max length of common prefix) even if must say is over abundant ; an analysis of nibbled key lengths in trie show this values

```
 Size    Count
 ---- --------
    0        1
    1       16
    2      256
    3     4096
    4    65536
    5  1048576
    6 15197337
    7  1563721
    8     8337
    9       38
   10        1
```